v4l2.go

```diff
@@ -418,28 +457,52 @@ func waitForFrame(fd uintptr, timeout uint32) (count int, err error) {
 
 }
 
+func getControl(fd uintptr, id uint32) (int32, error) {
+	ctrl := &v4l2_control{}
+	ctrl.id = id
+	err := ioctl.Ioctl(fd, VIDIOC_G_CTRL, uintptr(unsafe.Pointer(ctrl)))
+	return ctrl.value, err
+}
+
 func setControl(fd uintptr, id uint32, val int32) error {
 	ctrl := &v4l2_control{}
 	ctrl.id = id
 	ctrl.value = val
 	return ioctl.Ioctl(fd, VIDIOC_S_CTRL, uintptr(unsafe.Pointer(ctrl)))
 }
 
-func getControls(fd uintptr) map[uint32]string {
-	query := &v4l2_queryctrl{}
-	var controls map[uint32]string
+func queryControls(fd uintptr) map[string]control {
+	var controls map[string]control = make(map[string]control)
 	var err error
-	for query.id = V4L2_CID_BASE; err == nil; query.id++ {
-		err = ioctl.Ioctl(fd, VIDIOC_QUERYCTRL, uintptr(unsafe.Pointer(query)))
-		if err == nil {
-			controls[query.id] = CToGoString(query.name[:])
-		}
-	}
-	err = nil
-	for query.id = V4L2_CID_PRIVATE_BASE; err == nil; query.id++ {
+	// Don't use V42L_CID_BASE since it is the same as brightness.
+	var id uint32
+	for err != ioctl.ErrEINVAL {
+		id |= V4L2_CTRL_FLAG_NEXT_CTRL
+		query := &v4l2_queryctrl{}
+		query.id = id
 		err = ioctl.Ioctl(fd, VIDIOC_QUERYCTRL, uintptr(unsafe.Pointer(query)))
+		id = query.id
 		if err == nil {
-			controls[query.id] = CToGoString(query.name[:])
+			if (query.flags & V4L2_CTRL_FLAG_DISABLED) != 0 {
+				continue
+			}
+			var c control
+			switch query._type {
+			default:
+				continue
+			case V4L2_CTRL_TYPE_INTEGER, V4L2_CTRL_TYPE_INTEGER64:
+				c.c_type = c_int
+			case V4L2_CTRL_TYPE_BOOLEAN:
+				c.c_type = c_bool
+			case V4L2_CTRL_TYPE_MENU:
+				c.c_type = c_menu
+			}
+			c.id = id
+			c.min = query.minimum
+			c.max = query.maximum
+			// Normalise name (' ' -> '_', make lower case).
+			n := strings.Replace(strings.ToLower(CToGoString(query.name[:])), " ", "_", -1)
```

Comment:
we shouldn't do any string manipulations, nor should we work with control names as identifiers for controls.
Here are few reasons why:
- Control name comes from webcam driver. Nowhere is stated that names should be unique.
- User might want to display control name in UI, what should he do? Denormalize control name back?

Reply:
OK - my intent was to match what the existing v4l2 utils do (e.g v4l2-ctl -L)